Optimise HtmlEncode and HtmlDecode to produce less garbage

benjamin-hodgson opened the pull request:
This fixes https://github.com/dotnet/corefx/issues/13893. I'd noticed in a project of mine that the `TextWriter` overload of `WebUtility.HtmlEncode` was unexpectedly generating garbage.

I added a `System.Buffers` project reference to the `System.Runtime.Extensions` csproj and wrote a simple implementation of `StringBuilder` based on `ArrayPool`, as suggested by @jkotas in https://github.com/dotnet/corefx/pull/13875.

I benchmarked this implementation and while it appears to produce no garbage (except for the `string` that it returns) it does seem to run around 10-20% slower. I'm keen to get some input from you guys on what might be going on here.


Benchmark code (using BenchmarkDotNet):

```csharp
[MemoryDiagnoser]
public class Bench
{
    public static readonly string _toEncode = string.Concat(Enumerable.Repeat("<>\"'", 100));

    [Benchmark(Baseline = true)]
    public void String_Old()
    {
        WebUtility.HtmlEncode(_toEncode);
    }

    [Benchmark]
    public void String_New()
    {
        WebUtility.HtmlEncodeNew(_toEncode);
    }

    [Benchmark]
    public void TextWriter_Old()
    {
        using (var writer = TextWriter.Null)
        {
            WebUtility.HtmlEncode(_toEncode, writer);
        }
    }

    [Benchmark]
    public void TextWriter_New()
    {
        using (var writer = TextWriter.Null)
        {
            WebUtility.HtmlEncodeNew(_toEncode, writer);
        }
    }
}
```

And the results:

```
         Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |  Gen 0 |  Gen 1 | Allocated |
--------------- |---------:|----------:|----------:|-------:|---------:|-------:|-------:|----------:|
     String_Old | 5.394 us | 0.0987 us | 0.0924 us |   1.00 |     0.00 | 2.0828 | 0.0229 |   13120 B |
     String_New | 6.571 us | 0.1101 us | 0.1030 us |   1.22 |     0.03 | 0.7629 |      - |    4832 B |
 TextWriter_Old | 5.431 us | 0.1078 us | 0.1153 us |   1.01 |     0.03 | 2.0828 | 0.0229 |   13120 B |
 TextWriter_New | 6.080 us | 0.1041 us | 0.0974 us |   1.13 |     0.03 |      - |      - |       0 B |
```

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -45,20 +47,56 @@ public static string HtmlEncode(string value)
-            HtmlEncode(value, index, sb);
-            return StringBuilderCache.GetStringAndRelease(sb);
+            var sb = new PooledStringBuilder(value.Length);
+            try
```
You do not need try/catch. It is ok to turn the pooled into garbage on exceptions.

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -689,6 +727,68 @@ private static bool StringRequiresHtmlDecoding(string s)
 
 #endregion
 
+        private struct PooledStringBuilder
```
We have a helper class like this here: https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs that has a few more optimizations. Try using that instead.

benjamin-hodgson replied:
It looks like `ValueStringBuilder` is backed by a `Span` - what's the best way of writing a `Span<char>` to a `TextWriter`? I don't want to just loop over the span and call `Write` because that'll be slow.

jkotas replied:
`void Write(System.ReadOnlySpan<char> buffer)` method.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -45,20 +47,56 @@ public static string HtmlEncode(string value)
                 return value;
             }
 
-            StringBuilder sb = StringBuilderCache.Acquire(value.Length);
```
Is the project still using StringBuilderCache after this change?  If not, it can be removed from the .csproj.  If yes, should those places be changed to ValueStringBuilder as well?

benjamin-hodgson replied:
It's used elsewhere ([eg](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/src/System/ApplicationId.cs#L44)). I'd be happy to fix those other usages too, in a separate PR, if you like.

leotsarev replied:
Hi! Other usages have been removed in #30072

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -45,20 +47,56 @@ public static string HtmlEncode(string value)
+            {
+                output.Write(value);
+                return;
+            }
```
How valuable is this check? In cases where we do have something to encode, we end up paying for the extra checking.

benjamin-hodgson replied:
I just copied this from the existing `string` overload of `HtmlEncode` above. My gut tells me that the case when `value` doesn't have any encoding characters is quite common, probably more common than when it does. Plus `index` is used to copy the prefix of the string quickly, see [line 106](https://github.com/dotnet/corefx/pull/27250/files/67f6909e4581843ecfa2b5ab23050a17edaeb4d9#diff-831332bbb389db07350457a4d92095f8R106)

davidfowl commented:
FYI there's a much more efficient HtmlEncoder implementation here https://github.com/dotnet/corefx/blob/01fa16ffd618846a913cad719269c8bb441ceb28/src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs.

jkotas commented on src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs:
```diff
@@ -61,6 +68,12 @@ public bool TryCopyTo(Span<char> destination, out int charsWritten)
             }
         }
 
+        public void WriteTo(TextWriter writer)
```
This method is very specialized. It should rather be `.AsReadOnlySpan()` property. It will handle all future uses that need to get the string as Span.

benjamin-hodgson replied:
Are you sure? That sounds like it could be a bit risky: when the underlying array gets resized the old one gets returned to the pool (but the `ReadOnlySpan` will still have a reference to it).

jkotas replied:
This is internal type, not a public API. There are many ways you can use it in a wrong way. For example, you can accidentally pass it by value and create a second copy of it to introduce the same problem as if you use `AsReadOnlySpan` incorrectly by accident.

benjamin-hodgson replied:
[Done](https://github.com/dotnet/corefx/pull/27250/commits/ef33e71833b6b9d2ab03ed2c97cf3199e9bcf857)

benjamin-hodgson commented:
@davidfowl Thanks, I didn't know about `HtmlEncoder`.

jkotas commented on src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs:
```diff
@@ -185,7 +197,7 @@ private void Grow(int requiredAdditionalCapacity)
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void Clear()
+        public void Clear()
```
Nit: It may be better to call it `Dispose` now that it is public.

danmosemsft replied:
I assume going further and adding IDisposable would be too heavyweight for this -- especially since "leaking" is not a big deal.

jkotas replied:
Right. Also, interfaces do not really make sense for ref types - you cannot cast the ref type to the interface.

benjamin-hodgson replied:
A tangent: I haven't tried this, but I imagine a ref struct with an interface might have some limited use in the context of constrained generics. That is, if you had

```csharp
ref struct Foo : IBar {}
void Baz<T>(T baz) where T : IBar
```

you should be able to write

```csharp
Baz(new Foo());
```

benjamin-hodgson replied:
Although: I guess you're probably not allowed to use a ref struct as a type argument, because you don't know whether the method will use the `T` in a way that you can't use a ref struct (eg casting it to `object`).

jkotas commented on src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj:
```diff
@@ -250,6 +253,7 @@
     <ProjectReference Include="..\..\System.Diagnostics.Debug\src\System.Diagnostics.Debug.csproj" />
     <ProjectReference Include="..\..\System.Security.Principal\src\System.Security.Principal.csproj" />
     <ProjectReference Include="..\..\System.Runtime\src\System.Runtime.csproj" />
+    <ProjectReference Include="..\..\System.Buffers\src\System.Buffers.csproj" />
```
Do you actually need to add this reference? Span is in `System.Runtime` for netcoreapp - you do not need to reference `System.Buffers` to get it.

benjamin-hodgson replied:
Ah, good catch, thanks! I needed in an earlier commit, because I was using `ArrayPool` directly, but that's been removed now.

danmosemsft commented on src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs:
```diff
@@ -20,6 +21,12 @@ public ValueStringBuilder(Span<char> initialBuffer)
             _chars = initialBuffer;
             _pos = 0;
         }
+        public ValueStringBuilder(int initialCapacity)
```
Can you please add some test(s) to ValueStringBuilderTests for your new functionality?

danmosemsft commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -148,27 +173,44 @@ public static string HtmlDecode(string value)
+
+            if (string.IsNullOrEmpty(value))
+            {
+                output.Write(value);
```
This will presumably throw if value is null. But, I see previously it would have also thrown doing `value.Length`.

danmosemsft replied:
Seems we should probably throw ANE if value is null.

benjamin-hodgson replied:
Good spot, will do.

benjamin-hodgson replied:
I'm now throwing an `ArgumentNullException`, but I'm concerned that this is technically a breaking change (see [my comment](https://github.com/dotnet/corefx/pull/27250#issuecomment-371223676)). What do you think?

danmosemsft replied:
We are generally OK with changing NRE->ANE.

danmosemsft replied:
Oh I see you changed back. That's fine too.

danmosemsft commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -148,27 +173,44 @@ public static string HtmlDecode(string value)
+                return;
+            }
+
+            if (!StringRequiresHtmlDecoding(value))
```
The code was not previously doing this check. If we have to do it, it might be nice for StringRequiresHtmlDecoding to return the index of the first character to decode, so HtmlDecode can copy up to that point verbatim, the same optimization happening with IndexOfHtmlEncodingChars.
Without that, I am not sure whether this will make it slower or not.

benjamin-hodgson replied:
It was doing this check before, in the `HtmlDecode(string value)` method (which was being called by this one).

ahsonkhan commented on src/Common/tests/Tests/System/Text/ValueStringBuilderTests.cs:
```diff
@@ -152,6 +163,27 @@ public void Insert_IntCharInt_MatchesStringBuilder()
+                vsb.Append(s);
+            }
+
+            var resultSpan = vsb.AsReadOnlySpan();
```
We renamed AsReadOnlySpan to AsSpan.

Also, nit: we don't use resultSpan anywhere. Change to:
`var resultString = new string(vsb.AsSpan());`

benjamin-hodgson replied:
[Done](https://github.com/dotnet/corefx/pull/27250/commits/b1b17d2a0e58e863e86440651b188ac3aab86c77), thanks!

ahsonkhan commented on src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs:
```diff
@@ -20,6 +21,12 @@ public ValueStringBuilder(Span<char> initialBuffer)
             _chars = initialBuffer;
             _pos = 0;
         }
+        public ValueStringBuilder(int initialCapacity)
```
nit: add new line between the ctors

benjamin-hodgson replied:
[Done](https://github.com/dotnet/corefx/pull/27250/commits/b1b17d2a0e58e863e86440651b188ac3aab86c77#diff-1a641bcf15d0f95ae2b8a25f157ecd51R24), thanks!

jkotas commented on src/Common/src/CoreLib/System/Text/ValueStringBuilder.cs:
```diff
@@ -4,6 +4,7 @@
 
 using System.Buffers;
 using System.Diagnostics;
+using System.IO;
```
This `using` should not be needed anymore.

benjamin-hodgson replied:
[Done](https://github.com/dotnet/corefx/pull/27250/commits/b1b17d2a0e58e863e86440651b188ac3aab86c77#diff-1a641bcf15d0f95ae2b8a25f157ecd51L7), thanks

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -148,27 +176,46 @@ public static string HtmlDecode(string value)
+            sb.Dispose();
         }
 
         [SuppressMessage("Microsoft.Usage", "CA1806:DoNotIgnoreMethodResults", MessageId = "System.UInt16.TryParse(System.String,System.Globalization.NumberStyles,System.IFormatProvider,System.UInt16@)", Justification = "UInt16.TryParse guarantees that result is zero if the parse fails.")]
```
Since you are optimizing allocations, you can also replace `uint.TryParse(value.Substring` below with `uint.TryParse(value.AsSpan` below; and delete this SuppressMessage - it does not seem to be relevant anymore.

benjamin-hodgson replied:
[Done](https://github.com/dotnet/corefx/pull/27250/commits/b1b17d2a0e58e863e86440651b188ac3aab86c77#diff-831332bbb389db07350457a4d92095f8L246), good spot, thanks!

benjamin-hodgson commented:
Rebased. Apologies for the delay in getting this back up to speed.

The tests were failing because I changed the code to throw `ArgumentNullException` when the input string was null, per [@danmosemsft's suggestion](https://github.com/dotnet/corefx/pull/27250#discussion_r169736385). Previously `HtmlEncode`/`HtmlDecode` would return `null` (or write nothing to the `TextWriter`, depending on which overload you called) if the input was `null`. For now [I've updated the tests](https://github.com/dotnet/corefx/pull/27250/commits/36aeacdfdf3ff736456fd161a5eeb0f4d0af55d0) to match the new behaviour, but this is technically a breaking change; it's not unlikely that client code is relying on being able to pass `null` to that method. Do you think I should keep the new behaviour or put it back to how it was?

ahsonkhan commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -5,13 +5,15 @@
 using System.IO;
 using System.Runtime.CompilerServices;
 using System.Text;
+using System.Buffers;
```
nit: sort in alphabetical order

benjamin-hodgson replied:
I think this import isn't used any more actually. Good catch, thanks!

ahsonkhan commented:
@benjamin-hodgson, the System.Web.Tests.HttpUtilityTest/HtmlEncode_TextWriter(decoded: null, encoded: \"\") test is failing in CI 

https://mc.dot.net/#/user/benjamin-hodgson/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/127d2ab60cced49647237cd9c6774f5102fb1095/workItem/System.Web.HttpUtility.Tests/analysis/xunit/System.Web.Tests.HttpUtilityTest~2FHtmlEncode_TextWriter(decoded:%20null,%20encoded:%20%5C%22%5C%22)
```text
Message :
System.ArgumentNullException : Value cannot be null.
Parameter name: value
Stack Trace :
   at System.Net.WebUtility.HtmlEncode(String value) in D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\src\System\Net\WebUtility.cs:line 42
   at System.Web.Util.HttpEncoder.HtmlEncode(String value, TextWriter output) in D:\j\workspace\windows-TGrou---74aa877a\src\System.Web.HttpUtility\src\System\Web\Util\HttpEncoder.cs:line 139
   at System.Web.HttpUtility.HtmlEncode(String s, TextWriter output) in D:\j\workspace\windows-TGrou---74aa877a\src\System.Web.HttpUtility\src\System\Web\HttpUtility.cs:line 140
   at System.Web.Tests.HttpUtilityTest.HtmlEncode_TextWriter(String decoded, String encoded) in D:\j\workspace\windows-TGrou---74aa877a\src\System.Web.HttpUtility\tests\HttpUtility\HttpUtilityTest.cs:line 273
```

benjamin-hodgson commented:
@ahsonkhan Yep I know. I think I'm going to undo the `ArgumentNullException` change because it's a breaking change (see [my earlier comment](https://github.com/dotnet/corefx/pull/27250#issuecomment-371223676)). I haven't had the time yet though, will try and get to it over the weekend.

benjamin-hodgson commented:
Apologies for the delay. I've rebased, and erased the commits with the changed `ArgumentNullException` behaviour

ahsonkhan commented on src/Common/tests/Tests/System/Text/ValueStringBuilderTests.cs:
```diff
@@ -152,6 +163,26 @@ public void Insert_IntCharInt_MatchesStringBuilder()
+            }
+
+            var resultString = new string(vsb.AsSpan());
+            Assert.Equal(sb.ToString(), resultString);
```
nit: You could do something like this instead of using the string constructor and storing in a local:

Assert.Equal(sb.ToString(), vsb.AsSpan().ToString());

ahsonkhan commented:
@danmosemsft, do you have any other feedback?

danmosemsft commented:
Can you please share the latest perf data? Are we still 10-20% slower?

stephentoub commented:
With the changes made in https://github.com/dotnet/corefx/pull/29402, is this PR still relevant, or are parts of it?  Assuming it is. it'll need to be rebased, and then it'll be good to see perf data.

benjamin-hodgson commented:
@stephentoub I think it is still relevant, because the two PRs address different classes of allocations. I'll rebase this branch and get you some benchmarks next week

benjamin-hodgson commented:
I measured `HtmlEncode` and `HtmlDecode`, with an input string of 100k `<`s and 100k `&lt;`s respectively.

```
         Method | Categories |     Mean |     Error |    StdDev | Scaled | ScaledSD |    Gen 0 |    Gen 1 |    Gen 2 |  Allocated |
--------------- |----------- |---------:|----------:|----------:|-------:|---------:|---------:|---------:|---------:|-----------:|
 HtmlEncode_Old |     Encode | 1.143 ms | 0.0227 ms | 0.0360 ms |   1.00 |     0.00 | 343.7500 | 330.0781 | 248.0469 | 1573.09 KB |
 HtmlEncode_New |     Encode | 2.319 ms | 0.0340 ms | 0.0318 ms |   2.03 |     0.07 | 281.2500 | 281.2500 | 281.2500 | 1109.29 KB |
                |            |          |           |           |        |          |          |          |          |            |
 HtmlDecode_Old |     Decode | 5.274 ms | 0.0712 ms | 0.0666 ms |   1.00 |     0.00 | 242.1875 | 242.1875 | 242.1875 |  976.73 KB |
 HtmlDecode_New |     Decode | 4.924 ms | 0.0305 ms | 0.0255 ms |   0.93 |     0.01 |  54.6875 |  54.6875 |  54.6875 |  203.34 KB |
```

So allocations are down across the board and `HtmlDecode` is about the same speed, but `HtmlEncode` is significantly slower. (This is using the Core 2.1 release candidate on OSX.)

I think this is down to the fact that `HtmlEncode` appends a few characters at a time to the output, whereas `HtmlDecode` only appends one character at a time in this benchmark. `ValueStringBuilder` seems to be comparatively slow in the former case. I opened a separate issue to look into this: #29921

benjamin-hodgson commented:
I don't know why the build is broken. Appears to be only one platform, maybe it's just being flaky?

danmosemsft commented:
@benjamin-hodgson if you follow the "details" links you will see its' an unrelated (flaky) test failure. You can rerun like this:
@dotnet-bot test Windows x64 Debug Build please

danmosemsft commented:
@stephentoub any other comments

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -45,20 +46,43 @@ public static string HtmlEncode(string value)
+
+            // a thought for possible future optimisation: rather than build the whole output string,
+            // build 4kb-sized chunks and dump them into the writer as you go
+            // Of course would only make a difference for large strings
```
I expect in many cases the TextWriter that's passed in does its own buffering, such that if we were to skip the ValueStringBuilder and just issue writes to it, that's essentially what we'd be doing.

FWIW, I don't think we need this comment.  If this is something to be investigated, it'd be better as an issue in the repo.

danmosemsft replied:
+1 it's also more likely that someone would do this investigation if it was moved to an issue.

stephentoub commented:
> any other comments

I wouldn't want to merge this until the throughput regression is addressed.  It's great that allocation is reduced, but not at the expense of a 100% regression in throughput.

ViktorHofer commented:
I benchmarked with these commits. See results below. Benchmark file: https://gist.github.com/ViktorHofer/d296de07c6178e586a0fb4f3f343e440

## Before

|                    Method |     Mean |     Error |    StdDev |       Gen 0 | Allocated |
|-------------------------- |---------:|----------:|----------:|------------:|----------:|
|              EncodeString | 76.35 ms | 0.6652 ms | 0.5897 ms | 100125.0000 |  200.5 MB |
| EncodeStringAndTextWriter | 77.63 ms | 1.2074 ms | 1.1294 ms | 100125.0000 |  200.5 MB |
|              DecodeString | 78.74 ms | 0.5254 ms | 0.4658 ms |   8937.5000 |  17.91 MB |
| DecodeStringAndTextWriter | 78.73 ms | 0.6507 ms | 0.5768 ms |   8937.5000 |  17.91 MB |

## After

|                    Method |     Mean |     Error |    StdDev |      Gen 0 |  Allocated |
|-------------------------- |---------:|----------:|----------:|-----------:|-----------:|
|              EncodeString | 79.47 ms | 1.8111 ms | 2.2243 ms | 36375.0000 | 76480000 B |
| EncodeStringAndTextWriter | 76.84 ms | 0.7241 ms | 0.5653 ms |          - |        0 B |
|              DecodeString | 79.44 ms | 1.0369 ms | 0.9192 ms |  1562.5000 |  3296000 B |
| DecodeStringAndTextWriter | 79.58 ms | 1.1509 ms | 1.0203 ms |          - |        0 B |

ViktorHofer commented:
@benjamin-hodgson I rebased onto master and removed the StringBuilderCache reference in S.R.Extensions after it was removed from all other places with yesterdays change.

ViktorHofer commented:
I'm looking into further optimizations in the WebUtility class. Please stand by.

benjamin-hodgson commented:
@ViktorHofer Ah thanks very much for looking into that. Those results look really good - is the improvement due to https://github.com/dotnet/coreclr/issues/18005 ? If so should #29921 be closed now?

ViktorHofer commented:
The latest commit removes another string allocation and removes unnecessary unsafe code where boundary checks are eliminated by the CLR.

I noticed that the test string wasn't really very expressive and changed it to something which hits all the different paths in the implementation: `Hello! '\"<&>\u2665\u00E7 World \u0021\u0023\u003D\u003F <>\"\\&`. **Unfortunately with this test string I see the big regression @benjamin-hodgson talked about**. See below...

## Before

|                    Method |     Mean |     Error |    StdDev |      Gen 0 | Allocated |
|-------------------------- |---------:|----------:|----------:|-----------:|----------:|
|              EncodeString | 49.16 ms | 0.9494 ms | 1.1659 ms | 19062.5000 |  38.15 MB |
| EncodeStringAndTextWriter | 53.59 ms | 0.3883 ms | 0.3442 ms | 19062.5000 |  38.15 MB |


## After

|                    Method |     Mean |     Error |    StdDev |      Gen 0 |  Allocated |
|-------------------------- |---------:|----------:|----------:|-----------:|-----------:|
|              EncodeString | 75.99 ms | 0.3655 ms | 0.3052 ms | 16000.0000 | 33600000 B |
| EncodeStringAndTextWriter | 97.22 ms | 1.0154 ms | 0.8479 ms |          - |        0 B |

ahsonkhan commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -63,100 +70,99 @@ public static void HtmlEncode(string value, TextWriter output)
+                        output.Append("&#");
+
+                        // Use the buffer directly and reserve a conservative estimate of 10 chars.
+                        Span<char> s = output.AppendSpan(10);
```
nit: avoid single character variable names.

ahsonkhan commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -229,16 +235,9 @@ private static void HtmlDecode(string value, ref ValueStringBuilder output)
-                            {
-                                parsedSuccessfully = uint.TryParse(value.AsSpan(entityOffset + 1, entityLength - 1), NumberStyles.Integer, CultureInfo.InvariantCulture, out parsedValue);
-                            }
+                            bool parsedSuccessfully = value[entityOffset + 1] == 'x' || value[entityOffset + 1] == 'X'
```
This is difficult to parse and read. Can we cache some of these in a local? Like `value[entityOffset + 1]`?

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -38,104 +39,130 @@ public static string HtmlEncode(string value)
+            var sb = new ValueStringBuilder(value.Length);
+
+            // Prepend non encoding part
+            ReadOnlySpan<char> prefixSpan = value.AsSpan(0, index);
```
Nit: I think it would look better without an extra local:

```
// Prepend non encoding part
sb.Append(value.AsSpan(0, index));
```

jkotas commented:
> big regression

What do the numbers look like for Decoding?

ValueStringBuilder that is created for encoding is guaranteed to have insufficient size because of encoding always grows. Try changing the line that creates the ValueStringBuilder to `var sb = new ValueStringBuilder(value.Length + 200)` and see whether it makes any difference.

ViktorHofer commented:
> ValueStringBuilder that is created for encoding is guaranteed to have insufficient size (modulo bucketing done by ArrayPool) because of encoding always grows. Try changing the line that creates the ValueStringBuilder to var sb = new ValueStringBuilder(value.Length + 200) and see whether it makes any difference.

15ms average improvement! Nice!

ViktorHofer commented:
With @jkotas suggestion of increasing the buffer size (13ms) and switching to stackalloc for small inputs (12ms) I was able to make it faster than it currently is in master.

## Before

|                    Method |     Mean |     Error |    StdDev |      Gen 0 | Allocated |
|-------------------------- |---------:|----------:|----------:|-----------:|----------:|
|              EncodeString | 51.74 ms | 0.8133 ms | 0.6791 ms | 19062.5000 |  38.15 MB |
| EncodeStringAndTextWriter | 52.68 ms | 1.0402 ms | 1.3156 ms | 19062.5000 |  38.15 MB |
|              DecodeString | 83.84 ms | 1.4484 ms | 1.3549 ms |  5812.5000 |  11.75 MB |
| DecodeStringAndTextWriter | 86.87 ms | 1.5921 ms | 1.4892 ms |  5812.5000 |  11.75 MB |

## After

|                    Method |     Mean |     Error |    StdDev |      Gen 0 |  Allocated |
|-------------------------- |---------:|----------:|----------:|-----------:|-----------:|
|              EncodeString | 44.29 ms | 0.5136 ms | 0.4289 ms | 16000.0000 | 33600000 B |
| EncodeStringAndTextWriter | 50.87 ms | 0.9949 ms | 0.9772 ms |          - |        0 B |
|              DecodeString | 90.22 ms | 1.7041 ms | 1.5940 ms |  5812.5000 | 12320000 B |
| DecodeStringAndTextWriter | 94.55 ms | 1.8487 ms | 1.8157 ms |          - |        0 B |

**Note: I haven't yet looked at Decode at all...**

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -148,27 +191,67 @@ public static string HtmlDecode(string value)
-            HtmlDecode(value, sb);
-            return StringBuilderCache.GetStringAndRelease(sb);
+            char[] rawBuffer = null;
+            // We expect the encoded pattern to be at least twice as large as the decoded one.
```
I think that this is wrong assumption. I think that there are going to be just a very few escaped characters in the typical case.

ViktorHofer replied:
okay, I will trim the check down to 256. do you think the 80 for the encode check is conservative enough or should we maybe even increase it to 100 / 128?

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -148,27 +191,65 @@ public static string HtmlDecode(string value)
-            HtmlDecode(value, sb);
-            return StringBuilderCache.GetStringAndRelease(sb);
+            char[] rawBuffer = null;
+            Span<char> buffer = value.Length < 256 ?
```
This can be `value.Length <= 256`

ViktorHofer commented:
Should we remove the ValueStringBuilder overload with its tests now that we are not using it anymore or keep it as it might bring value for future scenarios?

jkotas commented:
> Should we remove the ValueStringBuilder ctor overload with its tests now that we are not using it anymore 

I think so.

ViktorHofer commented:
Thanks @benjamin-hodgson for your work, I'm done my changes here now. Both throughput and allocation figures are now better than on master therefore I believe this PR is now in a mergeable state. Thanks @jkotas and others for hints.

@stephentoub it would be great if you could glance over the changes before we merge this. thanks!

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -29,113 +31,152 @@ public static class WebUtility
+            char[] rawBuffer = null;
+            Span<char> buffer = value.Length < 80 ?
+                stackalloc char[256] :
+                (rawBuffer = ArrayPool<char>.Shared.Rent(value.Length + 200));
```
A comment on how these constants were chosen would be helpful.

ViktorHofer replied:
I added comments to explain why we chose these values. The fixed 200 additional characters could/should be transformed into a relative value but for that we first need more information on the distribution of input strings for HtmlEncode & HtmlDecode. See offline thread with Lee.

ViktorHofer replied:
I want to merge this in its current state as getting data might take some time. We can refine the buffer sizes afterwards.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -29,113 +31,152 @@ public static class WebUtility
+            if (rawBuffer != null)
+            {
+                ArrayPool<char>.Shared.Return(rawBuffer);
+            }
```
Why are we doing it this way rather than the previous approach of passing the initial capacity into VSB and having it do the ArrayPool access?  With this approach, we're going to hold onto this ArrayPool buffer no matter how many times the VSB grows, whereas if the VSB did it, it would return the original buffer to the pool when it first grew out of it.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -29,113 +31,152 @@ public static class WebUtility
+            Span<char> buffer = value.Length < 80 ?
+                stackalloc char[256] :
+                (rawBuffer = ArrayPool<char>.Shared.Rent(value.Length + 200));
+            var sb = new ValueStringBuilder(buffer);
```
Same comments/questions as above.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -29,113 +31,152 @@ public static class WebUtility
+                        // Use the buffer directly and reserve a conservative estimate of 10 chars.
+                        Span<char> encodingBuffer = output.AppendSpan(10);
+                        valueToEncode.TryFormat(encodingBuffer, out int charsWritten, provider: CultureInfo.InvariantCulture);
+                        output.Length -= (10 - charsWritten);
```
I think it'd be helpful to put the 10 that's used multiple times into a const that describes how it was picked, e.g. `const int MaxInt32Digits = 10` or something like that.  I assume valueToEncode can't be negative at this point?  If it could, max digits would need to be 11.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -29,113 +31,152 @@ public static class WebUtility
+
+                        // Use the buffer directly and reserve a conservative estimate of 10 chars.
+                        Span<char> encodingBuffer = output.AppendSpan(10);
+                        valueToEncode.TryFormat(encodingBuffer, out int charsWritten, provider: CultureInfo.InvariantCulture);
```
It shouldn't be necessary to specify the culture here, and doing so will make it slower.

ViktorHofer commented:
@benjamin-hodgson I squashed yours commits together so that the PR is ready to be merged, *not squashed*.

benjamin-hodgson commented:
@ViktorHofer OK, no problem. Thanks!
